From 24e8b632c87ad520a1dd41e2f269e23958d60a8c Mon Sep 17 00:00:00 2001
From: Erez Shaul <eshaul@habana.ai>
Date: Mon, 02 Sep 2024 17:57:39 +0300
Subject: [PATCH] [FSW-2067] fs1: phy fw prepare & copy - trigger fsm

-trigger pcie fw fsm to copy pcie sram locations and
pcie csr locations to the phy
-wait for fsm_phy_init_done

Change-Id: I39824d05ef044d246bb052d7c643f3303739cc94
Signed-off-by: Erez Shaul <eshaul@habana.ai>
---

diff --git a/arch/fs1/arc/common.c b/arch/fs1/arc/common.c
index 57a8ecd..d3307f7 100644
--- a/arch/fs1/arc/common.c
+++ b/arch/fs1/arc/common.c
@@ -540,6 +540,12 @@
 	return 0;
 }
 
+size_t get_csr_locations_data_size(void)
+{
+	/* this size is taken from: auto_gen_pcie_phy_csr_locations.h */
+	return csr_locations_size;
+}
+
 #define ALIGN_TO_16_BYTES(addr) (((addr) + 15) & ~0xF)
 #define PPBOOT_LAST_ADDR ((uint32_t) (&__SDATA_LOAD))
 
@@ -555,5 +561,5 @@
 		    (uint32_t)PHY_FW_SIZE, false);
 
 	/* Copy CSR_LOC_DATA to IO0_PCIE_FW_CSR_BASE */
-	memcpy((void *)IO0_PCIE_FW_CSR_BASE, (const void *)csr_loc_data, get_csr_locations_size());
+	memcpy((void *)IO0_PCIE_FW_CSR_BASE, (const void *)csr_loc_data, get_csr_locations_data_size());
 }
diff --git a/arch/fs1/arc/hl.h b/arch/fs1/arc/hl.h
index f6b54ed..2dfe0cf 100644
--- a/arch/fs1/arc/hl.h
+++ b/arch/fs1/arc/hl.h
@@ -75,5 +75,6 @@
 void hl_dev_psoc_pll_init(void);
 int fs1_mem_copy(uint64_t src, uint64_t dst, uint32_t size, bool is_hbw);
 void plat_pcie_drain_init(void);
+size_t get_csr_locations_data_size(void);
 
 #endif /* ARCH_FS1_ARC_HL_H_ */
diff --git a/arch/fs1/arc/palad.c b/arch/fs1/arc/palad.c
index 7e324ba..4cc38dc 100644
--- a/arch/fs1/arc/palad.c
+++ b/arch/fs1/arc/palad.c
@@ -6,6 +6,8 @@
 #include <psoc_pad_conf_regs.h>
 #include <pcie_dbi_regs.h>
 #include <pcie_dbi_masks.h>
+#include <pcie_fw_fsm_regs.h>
+#include <pcie_fw_fsm_masks.h>
 #include <pcie_aux_regs.h>
 #include <pcie_aux_masks.h>
 #include <psoc_boot_conf_regs.h>
@@ -22,7 +24,6 @@
 #include <fs1_efuse.h>
 #include <efuse.h>
 #include <hl.h>
-
 #include <address_space.h>
 
 #define SYS_CLOCK_FREQUENCY	250000000ULL /* PCI PLL DIV3 Emulation clk */
@@ -157,6 +158,69 @@
 	}
 }
 
+static void init_pcie_phy_fw_update(void)
+{
+	u32 val;
+
+	/*
+	 * The exact size of data written to SRAM_LOCATIONS and CSR_LOCATIONS
+	 * is written to PHY_FW_FSM_SIZE register.
+	 */
+	val = regmap_readl(mmIO_0_CFG_HALF_0_CXL_EP_PCIE_FW_FSM_BASE +
+			   mmPCIE_FW_FSM_PHY_FW_FSM_SIZE);
+	val &= ~PCIE_FW_FSM_PHY_FW_FSM_SIZE_PMA_SRAM_SIZE_M;
+	val |= PHY_FW_SIZE << PCIE_FW_FSM_PHY_FW_FSM_SIZE_PMA_SRAM_SIZE_S;
+	regmap_writel(val, mmIO_0_CFG_HALF_0_CXL_EP_PCIE_FW_FSM_BASE +
+			   mmPCIE_FW_FSM_PHY_FW_FSM_SIZE);
+
+	val = regmap_readl(mmIO_0_CFG_HALF_0_CXL_EP_PCIE_FW_FSM_BASE +
+			   mmPCIE_FW_FSM_PHY_FW_FSM_SIZE);
+	val &= ~PCIE_FW_FSM_PHY_FW_FSM_SIZE_PMA_CSR_SIZE_M;
+	val |= (uint16_t)get_csr_locations_data_size() << PCIE_FW_FSM_PHY_FW_FSM_SIZE_PMA_CSR_SIZE_S;
+	regmap_writel(val, mmIO_0_CFG_HALF_0_CXL_EP_PCIE_FW_FSM_BASE +
+			   mmPCIE_FW_FSM_PHY_FW_FSM_SIZE);
+
+	/*
+	 * PHY_FW_FSM_CTRL ➔ SRAM_ROM_SEL should be set to '0'
+	 * to disable default phy fw from ROM.
+	 */
+	val = regmap_readl(mmIO_0_CFG_HALF_0_CXL_EP_PCIE_FW_FSM_BASE +
+			   mmPCIE_FW_FSM_PHY_FW_FSM_CTRL);
+	val &= ~PCIE_FW_FSM_PHY_FW_FSM_CTRL_SRAM_ROM_SEL_M;
+	val |= 0 << PCIE_FW_FSM_PHY_FW_FSM_CTRL_SRAM_ROM_SEL_S;
+	regmap_writel(val, mmIO_0_CFG_HALF_0_CXL_EP_PCIE_FW_FSM_BASE +
+			   mmPCIE_FW_FSM_PHY_FW_FSM_CTRL);
+
+	/*
+	 * PHY_FW_FSM_CTRL ➔ FSM_START set to '1'
+	 * to start the phy fw loading to the fw memories.
+	 */
+	val = regmap_readl(mmIO_0_CFG_HALF_0_CXL_EP_PCIE_FW_FSM_BASE +
+			   mmPCIE_FW_FSM_PHY_FW_FSM_CTRL);
+	val &= ~PCIE_FW_FSM_PHY_FW_FSM_CTRL_FSM_START_M;
+	val |= 1 << PCIE_FW_FSM_PHY_FW_FSM_CTRL_FSM_START_S;
+	regmap_writel(val, mmIO_0_CFG_HALF_0_CXL_EP_PCIE_FW_FSM_BASE +
+			   mmPCIE_FW_FSM_PHY_FW_FSM_CTRL);
+
+	/*
+	 * FW should poll on PHY_FW_FSM_CTRL ➔ FW_FSM_PHY_INIT_DONE
+	 * which shows the completion of FW loading to the phy.
+	 */
+
+	/*
+	 * this code is commented out because pcie phy is not fully implemented
+	 * in palladium and it will hang the simulation.
+	 * once pcie phy is implemented, this code should be uncommented.
+	 * jira ticket: H15-352
+	 *
+	 * while (!(regmap_readl(mmIO_0_CFG_HALF_0_CXL_EP_PCIE_FW_FSM_BASE +
+	 * mmPCIE_FW_FSM_PHY_FW_FSM_CTRL) &
+	 * PCIE_FW_FSM_PHY_FW_FSM_CTRL_FW_FSM_PHY_INIT_DONE_M));
+	 *
+	 * preboot_puts("FW copied to phy successfully\n");
+	 */
+}
+
 #define BAR_SIZE_1GB		(SZ_1G - 1)
 #define VF_RES_BAR_SIZES	(PCIE_DBI_VF_RESBAR_CAP_REG_2_REG_V512GB_M | \
 				 PCIE_DBI_VF_RESBAR_CAP_REG_2_REG_V256GB_M | \
@@ -201,6 +265,8 @@
 
 void preboot_pcie_wait_for_link(void)
 {
+	init_pcie_phy_fw_update();
+
 	palad_pcie_params_init();
 
 	regmap_writel(LTTSM_ENA_MASK, mmIO_0_CFG_HALF_0_CXL_EP_PCIE_AUX_BASE +
